README
NAME
    Business::Address::POBox - Check whether an address looks like a P.O.Box

VERSION
    version 1.100820

SYNOPSIS
        use Business::Address::POBox;

        my $address = 'PF 34, 1010 Wien';
        if (Business::Address::POBox->new->is_pobox($address)) {
            # do something with the address
        }

DESCRIPTION
    This class tries to determine whether or not an string refers to a P.O.
    box. This is sometimes relevant if your business process, for legal
    reasons, needs a real address and not a P.O. box.

    It has predefined blacklists and whitelists that should catch most
    English and German P.O. box addresses, but you can modify these lists
    with the methods provided. Note that the entries are literal strings,
    not regular expressions.

METHODS
  new
        my $obj = Business::Address::POBox->new;
        my $obj = Business::Address::POBox->new(%args);

    Creates and returns a new object. The constructor will accept as
    arguments a list of pairs, from component name to initial value. For
    each pair, the named component is initialized by calling the method of
    the same name with the given value. If called with a single hash
    reference, it is dereferenced and its key/value pairs are set as
    described before.

  init
    Just calls "update()" in case the blacklist and/or whitelist was set
    during the "new()" call.

  update
    Call this method when you've changed the "whitelist()" or the
    "blacklist()" so the matcher knows about the changes.

  is_pobox
    This is the central method of this class. It takes a string argument and
    checks it against the whitelist and the blacklist. Returns a boolean
    value - true if the string passes the whitelist or is at least not
    caught by the blacklist, false if the string is caught by the blacklist.

  is_pobox_relaxed
    Like "is_pobox()", but once a string passes the whitelist, it is not
    checked against the blacklist anymore. That is, if a string matches the
    whitelist, it is valid. If not, it is checked against the blacklist - if
    it matches, it is invalid. If it matches neither whitelist nor
    blacklist, it is valid.

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Business-Address-POBox
    >.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <http://search.cpan.org/dist/Business-Address-POBox/>.

    The development version lives at
    <http://github.com/hanekomu/Business-Address-POBox/>. Instead of sending
    patches, please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2007 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Business-Address-POBox

1.100820  2010-03-23 19:35:36 Europe/Vienna
    - converted the distribution to Dist::Zilla-style

0.10 Mon May  4 16:29:37 CEST 2009 (Marcel Gruenauer <marcel@cpan.org>)
     - 'Alte Post' was falsely recognized as a POBox; fixed

0.09  Fri Dec 12 11:51:11 CET 2008 (Marcel Gruenauer <marcel@cpan.org>)
     - 'Am Steinkopf 17' was falsely recognized as a POBox; fixed

0.08  Thu Jul 25 15:27:43 CEST 2008 (Marcel Gruenauer <marcel@cpan.org>)
     - Changed dist style and Changes back to standard. People didn't like it -
       the nail that sticks out gets hammered down.
     - Added standard test files; this will also help with CPANTS.

0.07  Thu, 29 May 2008 19:49:25 -0000 (Marcel Gruenauer <marcel@cpan.org>)
     - set the version to 0.07
     - .shipit: fixed svk.tagpattern
     - .shipit: fixed typo in svk.tagpattern
     - added is_pobox_relaxed()

0.06  Thu, 29 May 2008 12:00:55 -0000 (Marcel Gruenauer <marcel@cpan.org>)
     - set the version to 0.06
     - patterns are now interpreted as literal text, not as regexes
     - moved generated documentation of interited methods to after
       documentation of this class' methods so the manpage is more readable
     - .shipit: fixed svk.tagpattern
     - Business::Address::POBox: now uses regexes instead of literal strings
     - t/01_is_pobox.t: added more tests

0.05  Sun, 25 Nov 2007 10:10:51 -0000 (Marcel Gruenauer <marcel@cpan.org>)
     - set the version to 0.05
     - Makefile.PL: Test::Exception now required, not build_required; some CPAN
       testers didn't pick it up

0.04  Fri, 23 Nov 2007 13:34:56 -0000 (Marcel Gruenauer <marcel@cpan.org>)
     - set the version to 0.04
     - Makefile.PL: build_requires Test::Exception

0.03  Wed, 21 Nov 2007 18:20:47 -0000 (Marcel Gruenauer <marcel@cpan.org>)
     - Added missing requirement for Class::Accessor::Constructor
     - .shipit: added Twitter step
     - Makefile.PL: added auto_install() and process_templates()
     - updated MANIFEST
     - updated MANIFEST
     - updated t/perlcriticrc
     - tags: BUGFIX

0.02  Mon, 12 Nov 2007 12:34:03 -0000 (Marcel Gruenauer <marcel@cpan.org>)
     - Fixed newlines in pod
     - tags: DOCFIX

0.01  Mon, 12 Nov 2007 12:28:06 -0000 (Marcel Gruenauer <marcel@cpan.org>)
     - original version


lib/Business/Address/POBox.pm
use 5.008;
use strict;
use warnings;

package Business::Address::POBox;
our $VERSION = '1.100820';

# ABSTRACT: Check whether an address looks like a P.O.Box
use String::BlackWhiteList;
use parent qw(Class::Accessor::Complex Class::Accessor::Constructor);
__PACKAGE__->mk_constructor->mk_object_accessors(
    'String::BlackWhiteList' => 'matcher')
  ->mk_array_accessors(qw(blacklist whitelist));
use constant DEFAULTS => (
    blacklist => [
        '\b(BOX|POB|POST(BOX|SCHACHTEL|FACH|LAGERND|BUS)?|POBOX)\b',
        '\b(P\.?\s*O\.?(\s*B(\.|OX))?)\b',
        '(^|\b)P\.?F\.?(-|\s+)\d',
    ],
    whitelist => [
        'Pf(-|\s+)\D',
'\b((Alte|An\s+der(\s+alten)?)\s+Post|Post(-|\s+)(Road|Rd|Street|St|Avenue|Av|Alley|Drive|Grove|Walk|Parkway|Row|Lane|Bridge|Boulevard|Square|Garden|Strasse|Gasse|Allee|Platz))\b',
    ],
);

sub init {
    my $self = shift;
    $self->update;
}

sub update {
    my $self = shift;
    for ($self->matcher) {
        $_->blacklist($self->blacklist);
        $_->whitelist($self->whitelist);
        $_->update;
    }
}

sub is_pobox {
    my ($self, $text) = @_;
    !$self->matcher->valid($text);
}

sub is_pobox_relaxed {
    my ($self, $text) = @_;
    !$self->matcher->valid_relaxed($text);
}
1;


__END__
=pod

=head1 NAME

Business::Address::POBox - Check whether an address looks like a P.O.Box

=head1 VERSION

version 1.100820

=head1 SYNOPSIS

    use Business::Address::POBox;

    my $address = 'PF 34, 1010 Wien';
    if (Business::Address::POBox->new->is_pobox($address)) {
        # do something with the address
    }

=head1 DESCRIPTION

This class tries to determine whether or not an string refers to a P.O. box.
This is sometimes relevant if your business process, for legal reasons, needs
a real address and not a P.O. box.

It has predefined blacklists and whitelists that should catch most English and
German P.O. box addresses, but you can modify these lists with the methods
provided. Note that the entries are literal strings, not regular expressions.

=head1 METHODS

=head2 new

    my $obj = Business::Address::POBox->new;
    my $obj = Business::Address::POBox->new(%args);

Creates and returns a new object. The constructor will accept as arguments a
list of pairs, from component name to initial value. For each pair, the named
component is initialized by calling the method of the same name with the given
value. If called with a single hash reference, it is dereferenced and its
key/value pairs are set as described before.

=head2 init

Just calls C<update()> in case the blacklist and/or whitelist was set during
the C<new()> call.

=head2 update

Call this method when you've changed the C<whitelist()> or the C<blacklist()>
so the matcher knows about the changes.

=head2 is_pobox

This is the central method of this class. It takes a string argument and
checks it against the whitelist and the blacklist. Returns a boolean value -
true if the string passes the whitelist or is at least not caught by the
blacklist, false if the string is caught by the blacklist.

=head2 is_pobox_relaxed

Like C<is_pobox()>, but once a string passes the whitelist, it is not checked
against the blacklist anymore. That is, if a string matches the whitelist, it
is valid. If not, it is checked against the blacklist - if it matches, it is
invalid. If it matches neither whitelist nor blacklist, it is valid.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Business-Address-POBox>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Business-Address-POBox/>.

The development version lives at
L<http://github.com/hanekomu/Business-Address-POBox/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.11;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'Business::Address::POBox',
                       'DISTNAME' => 'Business-Address-POBox',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.11'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {},
                       'ABSTRACT' => 'Check whether an address looks like a P.O.Box',
                       'EXE_FILES' => [],
                       'VERSION' => '1.100820',
                       'PREREQ_PM' => {
                                        'Class::Accessor::Complex' => '0',
                                        'Scalar::Util' => '0',
                                        'parent' => '0',
                                        'Test::Exception' => '0',
                                        'English' => '0',
                                        'constant' => '0',
                                        'Carp' => '0',
                                        'Test::More' => '0.88',
                                        'Class::Accessor::Constructor' => '0',
                                        'String::BlackWhiteList' => '0',
                                        'File::Find' => '0',
                                        'File::Temp' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{LICENSE}
  unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(%WriteMakefileArgs);



